### PR TITLE
Jl/caip multichain/verify scope method

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25589?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/lib/multichain-api/provider-request.js [hunk 1]
```diff
@@ -18,38 +11,45 @@ export async function providerRequestHandler(
...
+    request.origin,
+    Caip25EndowmentPermissionName,
+    Caip25CaveatType,
+  );
+  if (!caveat) {
     return end(new Error('missing CAIP-25 endowment'));
   }
 
-  const chainId = scope.split(':')[1];
+  // TODO: consider case when scope is defined in requireScopes and optionalScopes
```
Comment (adonesky1) [inline app/scripts/lib/multichain-api/provider-request.js:25]: 🙄 💯

**Review by shanejonas — APPROVED** [review]

File: app/scripts/lib/multichain-api/caip-25.ts [hunk 2]
```diff
@@ -141,3 +141,21 @@ export const isSupportedScopeString = (scopeString: string) => {
 
   return false;
 };
+
+export const flattenScope = (
```
Comment (adonesky1) [inline app/scripts/lib/multichain-api/caip-25.ts:145]: can we just add a jsdoc or comment indicating the formatting case (linking to CAIP-25) this is handling?

File: app/scripts/lib/multichain-api/provider-request.js [hunk 3]
```diff
@@ -18,38 +11,45 @@ export async function providerRequestHandler(
...
+    caveat.value.requiredScopes[scope] || caveat.value.optionalScopes[scope];
+
+  if (!scopeObject) {
+    return end(new Error('unauthorized (scopeObject missing)'));
+  }
 
-  if (!chainId) {
-    return end(new Error('missing chainId'));
+  if (!scopeObject.methods.includes(wrappedRequest.method)) {
+    return end(new Error('unauthorized (method missing in scopeObject)'));
```
Comment (adonesky1) [inline app/scripts/lib/multichain-api/provider-request.js:34]:
```suggestion
    return end(new Error('unauthorized (method not authorized)'));
```

File: app/scripts/lib/multichain-api/provider-request.js [hunk 4]
```diff
@@ -18,38 +11,45 @@ export async function providerRequestHandler(
...
   }
 
+  // Do we need to try catch this?
+  const { reference } = parseCaipChainId(scope);
```
Comment (adonesky1) [inline app/scripts/lib/multichain-api/provider-request.js:38]: yes for sure, unless we're validating the CAIP-2 scope somewhere above?

File: app/scripts/lib/multichain-api/provider-request.js (hunk 3 above)
Comment (jiexi) [reply app/scripts/lib/multichain-api/provider-request.js:34]: we were going to come back and decide on proper error codes and messages at a later point haha

File: app/scripts/lib/multichain-api/provider-request.js (hunk 4 above)
Comment (jiexi) [reply app/scripts/lib/multichain-api/provider-request.js:38]: ahh dumb question. When i wrote this i was thinking it wasn't coming from the request for some reason

File: app/scripts/lib/multichain-api/caip-25.ts (hunk 2 above)
Comment (jiexi) [reply app/scripts/lib/multichain-api/caip-25.ts:155]: added here https://github.com/MetaMask/metamask-extension/pull/25589/commits/936858ab0233cbe21f025cd11027b1fb422d1bc9

File: app/scripts/lib/multichain-api/provider-request.js (hunk 4 above)
Comment (jiexi) [reply app/scripts/lib/multichain-api/provider-request.js]: resolved here https://github.com/MetaMask/metamask-extension/pull/25589/commits/3c687d8fa11126338db1abdc7e5dc588180c6f5a

**Review by adonesky1 — APPROVED** [review]

**Pull request merged**